### PR TITLE
[Graphbolt] Add pickle serialization support for Graphbolt SampledSubgraph

### LINKED ISSUE #5953: [GraphBolt] GraphBolt objects cannot be serialized during multiprocessing
## 🐛 Bug

It seems that GraphBolt objects like `SampledSubgraph` cannot be passed around in multiprocessing due to lack of serialization method.

## To Reproduce

```python
import dgl.graphbolt
import torch
import multiprocessing as mp
import scipy.sparse as sp

def rand_csc_graph(N, density):
    adj = sp.random(N, N, density)
    adj = adj + adj.T
    adj = adj.tocsc()

    indptr = torch.LongTensor(adj.indptr)
    indices = torch.LongTensor(adj.indices)

    graph = dgl.graphbolt.from_csc(indptr, indices)

    return graph

def _entry(graph, q):
    adjs = []
    seeds = torch.arange(5)

    for hop in range(2):
        sg = graph.sample_neighbors(seeds, torch.LongTensor([2]))
        seeds = sg.indices
        adjs.insert(0, sg)

    print('Putting')
    q.put(adjs)
    print('Done put')

if __name__ == '__main__':
    graph = rand_csc_graph(200, 0.15)
    q = mp.Queue()
    proc = mp.Process(target=_entry, args=(graph, q))
    proc.start()
    print('Getting')
    item = q.get()
    print('Done get')
    proc.join()
    print(item)
```

This code snippet hangs indefinitely, either silently or throws the following error:
```
Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
RuntimeError: Tried to serialize object __torch__.torch.classes.graphbolt.SampledSubgraph which does not have a __getstate__ method defined!
```

## Expected behavior

The data should be successfully printed.  Replacing `q.put(adjs)` to something else like `q.put(42)` works.  So it should be something explicit to GraphBolt.

## Environment

 - DGL Version (e.g., 1.0): master
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):
 - OS (e.g., Linux):
 - How you installed DGL (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
Implemented `__setstate__` and `__getstate__` function for `SampledSubgraph` which enables pickle to dump it.
Resolve #5953.
[Updated] Added unit test.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- Modified `graphbolt/include/graphbolt/sampled_subgraph.h` and `graphbolt/src/sampled_subgraph.cc` to add `SetState()` and `GetState()` functions. It serializes the graph into a vector of tensors.
- Modified `graphbolt/src/python_bindings.cc`. Use `m.class_<>.def_pickle()` to bind them onto python `__setstate__` and `__getstate__` functions. (Ref: [pytorch docs](https://pytorch.org/tutorials/advanced/torch_script_custom_classes.html#defining-serialization-deserialization-methods-for-custom-c-classes))
- If in the future there is `Dict` type attributes, `SetState()` and `GetState()` could be easily modified to support it.
- Added unit test in `tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py`

### CODE REVIEW

File: graphbolt/include/graphbolt/sampled_subgraph.h [hunk 1]
```diff
@@ -113,6 +113,24 @@ struct SampledSubgraph : torch::CustomClassHolder {
    * subgraph.
    */
   torch::optional<torch::Tensor> type_per_edge;
+
+  /**
+   * @brief Version number to indicate graph version in serialization and
+   * deserialization.
+   */
+  static constexpr int64_t kSampledSubgraphSerializeVersionNumber = 1;
```
Comment (frozenbugs) [inline graphbolt/include/graphbolt/sampled_subgraph.h:121]: Does external class need to access this field?

File: graphbolt/src/sampled_subgraph.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+void SampledSubgraph::SetState(std::vector<torch::Tensor>& state) {
+  // Iterator.
+  uint32_t i = 0;
+
+  // Version number.
+  torch::Tensor version_num_tensor =
+      torch::ones(1, torch::TensorOptions().dtype(torch::kInt64)) *
+      SampledSubgraph::kSampledSubgraphSerializeVersionNumber;
+  TORCH_CHECK(
+      version_num_tensor.equal(state[i++]),
```
Comment (frozenbugs) [inline graphbolt/src/sampled_subgraph.cc:66]: Avoid change value in `check` statement, which is usually error-prone.

One option:
```
  TORCH_CHECK(
      version_num_tensor.equal(state[0),

  // Iterator.
  uint32_t i = 1;
```

File: graphbolt/include/graphbolt/sampled_subgraph.h (hunk 1 above)
Comment (RamonZhou) [reply graphbolt/include/graphbolt/sampled_subgraph.h:121]: No I think. Should it be set to private and non-static?

File: tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            adjs[hop].type_per_edge is None
+            and result[hop].type_per_edge is None
+        ) or torch.equal(adjs[hop].type_per_edge, result[hop].type_per_edge)
+
+
+@unittest.skipIf(
+    F._default_context_str == "gpu",
+    reason="Graph is CPU only at present.",
+)
+def test_subgraph_serialization():
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py:81]: What will happen if a new tensor added to sampled_subgraph class but the developer forget to add the tensor in the GetState and SetState?

File: tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    adjs = []
+    seeds = torch.arange(5)
+
+    # Sampling.
+    for hop in range(2):
+        sg = graph.sample_neighbors(seeds, torch.LongTensor([2]))
+        seeds = sg.indices
+        adjs.append(sg)
+
+    # 1. Put.
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py:43]: The 1, 2, 3, 4 which representing the order of the operators is not very clear, and also how do you guarantee the #2 happens after #1?

File: graphbolt/include/graphbolt/sampled_subgraph.h (hunk 1 above)
Comment (frozenbugs) [reply graphbolt/include/graphbolt/sampled_subgraph.h:121]: move it to sampled_subgraph.cc within the namespace should be fine.

File: tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        ) or torch.equal(adjs[hop].type_per_edge, result[hop].type_per_edge)
+
+
+@unittest.skipIf(
+    F._default_context_str == "gpu",
+    reason="Graph is CPU only at present.",
+)
+def test_subgraph_serialization():
+    # Create a sub-process.
+    q = mp.Queue()
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_c_sampled_subgraph.py:82]: rename q -> queue, avoid unnecessary abbreviation.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**